File: dongdaang/string/open_the_lock/Open_the_Lock.py
# https://leetcode.com/problems/open-the-lock/


###########돌아가긴 하는데 애매함(5900ms)############

from collections import deque

class Solution:
    def openLock(self, deadends, target):    
        if '0000' in deadends:
            return -1
        def bfs():
            global cnt
            while queue:
                n = len(queue)      #현재 큐의 크기
                for _ in range(n):              #현재 큐의 크기만큼 시도한 후 cnt 1 증가
                    now = queue.popleft()       #now -> 현재 시도하는 번호
                    if now == target:
                        return cnt
                    for i in range(4):          #4자리 모두 시도
                        for j in range(2):      # +1, -1 모두 시도
                            next = now[:i] + str((int(now[i]) + dx[j]) % 10) + now[i+1:]    #next -> 변경된 번호
                            if next in deadends or visited[int(next)]:
                                continue
                            queue.append(next)
                            visited[int(next)] = True
                cnt += 1
            return -1
        
        dx = [-1, 1]    #숫자 한칸씩 변경해주기 위한 변수
        global cnt      #변경 횟수 기록해주는 변수
        cnt = 0
        queue = deque(['0000'])     #변경 가능한 번호 담아주는 큐
        visited = [False] * 10000   #이미 방문한 번호를 표시해주는 배열
        visited[0] = True
        
        return bfs()

a = Solution()
print(a.openLock(["8887","8889","8878","8898","8788","8988","7888","9888"], '8888'))




###################솔루션(600ms)###################

class Solution:
    def openLock(self, deadends, target):
        dead_set = set(deadends)
        queue = deque([('0000', 0)])    #변경된 횟수를 해당 번호와 튜플로 묶어서 큐에 저장
        visited = set('0000')           #이미 방문한 번호 세트로 저장

        while queue:
            (string, steps) = queue.popleft()
            if string == target:
                return steps
            elif string in dead_set:
                continue
            for i in range(4):
                digit = int(string[i])
                for move in [-1, 1]:        #move -> 번호 변경을 위한 변수
                    new_digit = (digit + move) % 10
                    new_string = string[:i]+str(new_digit)+string[i+1:]     #new_string -> 변경된 번호
                    if new_string not in visited:
                        visited.add(new_string)
                        queue.append((new_string, steps+1))
        return -1

File: CokeLee777/src/com/leetcode/string/openthelock/OpenTheLock.java
package com.leetcode.string.openthelock;

import java.util.*;
import java.util.stream.Collectors;

/**
 * https://leetcode.com/problems/open-the-lock/
 */
public class OpenTheLock {

    public static class MySourceCode{

        private static Set<String> deads = new HashSet<>();
        private static Set<String> visited = new HashSet<>();

        public int openLock(String[] deadends, String target) {
            Queue<String> queue = new LinkedList<>();
            deads = Arrays.stream(deadends)
                    .collect(Collectors.toSet());
            visited = new HashSet<>();
            //시작위치 방문처리 및 큐에 삽입
            queue.offer("0000");
            visited.add("0000");

            //큐가 빌때까지 반복
            int step = 0;
            while(!queue.isEmpty()){
                //현재 번호에서 다음으로 가는 방법을 센다
                int size = queue.size();
                while(size-- > 0){
                    String now = queue.poll();
                    //deadends라면 무시
                    if(isDeadends(now)){
                        size--;
                        continue;
                    }
                    //아니라면 타겟이랑 비교
                    if(isTarget(target, now)) return step;
                    //현재 비밀번호에서 갈 수 있는 비밀번호 찾기
                    findPossibleWayFromNow(queue, now);
                }
                step++;
            }
            //비밀번호를 맞출수있는 방법이 없다면 -1출력
            return -1;
        }

        private void findPossibleWayFromNow(Queue<String> queue, String now) {
            //네 번호 모두 반복하면서 가능한 번호 삽입
            StringBuilder tmp = new StringBuilder(now);
            for(int i = 0; i < 4; i++){
                //i번째 위치에서 번호
                char x = tmp.charAt(i);
                //i번째 비밀번호만 1증가, 1감소
                String s1 = tmp.substring(0, i) + (x == '9' ? 0 : x - '0' + 1) + tmp.substring(i + 1);
                String s2 = tmp.substring(0, i) + (x == '0' ? 9 : x - '0' - 1) + tmp.substring(i + 1);
                //s1, s2가 방문하지 않았고, deadends가 아니면 큐에 삽입
                if(isNotVisited(s1) && !isDeadends(s1)){
                    queue.offer(s1);
                    visited.add(s1);
                }
                if(isNotVisited(s2) && !isDeadends(s2)){
                    queue.offer(s2);
                    visited.add(s2);
                }
            }
        }

        private boolean isNotVisited(String s1) {
            return !visited.contains(s1);
        }

        private boolean isDeadends(String now) {
            return deads.contains(now);
        }

        private boolean isTarget(String target, String now) {
            return now.equals(target);
        }
    }
}
